File: 2/2_cuatrimestre/PDOO/practicas/practica 4/deepSpaceRuby/Damage.rb
require_relative 'WeaponType'
require_relative 'DamageToUI'


module Deepspace
    class Damage
        attr_reader :nShields

        private_class_method def initialize(s)
            @nShields = s
        end
        
        def getUIversion
            raise "Método abstracto"
        end 

        def adjust(w, s)
            raise "Método abstracto"
        end

        def discardWeapon(w)
            raise "Método abstracto"
        end


        def discardShieldBooster
            if @nShields > 0
                @nShields -= 1
            end 
        end

        def hasNoEffect
            raise "Método abstracto"
        end

        def to_s
            raise "Método abstracto"
        end
    end
end

File: 2/2_cuatrimestre/PDOO/practicas/practica 4/deepSpace_4/src/deepspace/Damage.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package deepspace;

import java.util.ArrayList;


/**
 *
 * @author manbolq
 */


public abstract class Damage{
    protected int nShields;
    
    /**************************************************************************/
    /**************************    CONSTRUCTOR   ******************************/
    /**************************************************************************/
    
    Damage(int s){
        nShields = s;
    }
    
    
    /**************************************************************************/
    /***********************    METODOS ABSTRACTOS  ***************************/
    /**************************************************************************/
    
    public abstract Damage copy();
    
    abstract DamageToUI getUIversion();
    
    public abstract Damage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s);
    
    public abstract void discardWeapon(Weapon w);
    
    public abstract boolean hasNoEffect();
    
    @Override
    public abstract String toString();
    
    
    /**************************************************************************/
    /****************************    METODOS   ********************************/
    /**************************************************************************/   
    
    public void discardShieldBooster(){
        if (nShields > 0)
            nShields--;
    }
    
    
    public int getNShields(){
        return nShields;
    }
}